### ISSUE TITLE
How to combine `inc` and `set_message` so that `pos` and `msg` are in sync ?

### ISSUE THREAD

**Issue by malaire** [console-rs/indicatif#376]
Following code gives this output (using current main branch, and also in `0.16.2` so this behaviour hasn't changed):
```
0/3 Step 1
1/3 Step 2
2/3 Step 3
3/3 Step 3
```
If I move `progress.inc(1);` before `set_message` then I get this output instead:
```
1/3
2/3 Step 1
3/3 Step 2
3/3 Step 3
```
But what I want is to have `pos` and `msg` in sync like this:
```
1/3 Step 1
2/3 Step 2
3/3 Step 3
```

Code:
```
fn main() {
    let progress = ProgressBar::new(3);
    progress.set_style(ProgressStyle::default_bar().template("{pos}/{len} {msg}"));
    for n in 1..=3 {
	progress.set_message(format!("Step {}", n));
        progress.inc(1);
        std::thread::sleep(Duration::from_secs(2));
    }
    progress.finish();
}
```

**Comment by djc** [comment]
At what commit did you test main? I think maybe 4be36b94be144af5ac7461f080595ce7b42c45a5 fixed this case?

If not, I do think it would make sense for the code you've written here to do what you expect. Based on looking at current main (0c2287886c40362cda194db401abb5f109db68c9) it's not obvious to me why it wouldn't.

**Comment by malaire** [comment]
I just re-tested with current main (0c228788) and behavior is as I said above.

Adding `std::thread::sleep(Duration::from_millis(100));` between `set_message` and `inc` fixes this, so seems to be some kind of timing issue?

**Comment by malaire** [comment]
It seems to me that both `set_message`+`inc` and `inc`+`inc` examples cause two calls with nearly same timestamp to
`ProgressBar::update_and_draw`, which eventually calls `BarState::update` which returns `false` for second call
because of rate limit and so bar is not redrawn.

So getting back to first example:

```
fn main() {
    let progress = ProgressBar::new(3);
    progress.set_style(ProgressStyle::default_bar().template("{pos}/{len} {msg}"));
    for n in 1..=3 {
	progress.set_message(format!("Step {}", n));
        progress.inc(1);
        std::thread::sleep(Duration::from_secs(2));
    }
    progress.finish();
}
```

```
; message is set to "Step 1", bar is drawn
0/3 Step 1
; step is increased to 1, bar is not drawn
; --- sleep ---

; message is set to "Step 2", bar is drawn
1/3 Step 2
; step is increased to 2, bar is not drawn
; --- sleep ---

; message is set to "Step 3", bar is drawn
2/3 Step 3
; step is increased to 3, bar is not drawn
; --- sleep ---

; finish
3/3 Step 3
```

**Comment by malaire** [comment]
I can think of two possible fixes here:

Simple fix for immediate problem would be to add `inc_with_message` which both increments position and sets message in a single update.

More complex fix: When state is changed and draw can't be done immediately because of rate limit, future draw is scheduled at the earliest time allowed by rate limit. This requires separate thread for scheduled draw.

**Comment by djc** [comment]
For the `inc()` + `inc()` issue, I don't think I care. It's impossible to both (a) rate limit draws and (b) never show "incorrect" values in case your steps have weirdly distributed timing. If your actual use case has this kind of timing, I suggest you change the default draw limit to something that doesn't have this issue.

For the `inc()` + `message()` case, I definitely don't want to spawn a separate thread to draw in the future. So the two possible directions I see are:

1. Don't draw on message; only mutations that change the position get a chance of drawing
2. Add a separate method that allows users to mutate multiple parts of the state at the same time (your simple fix, generalized)

I think 2 would take the form of something like `pb.update(|state| { state.pos += 15; state.message = "Step 1".into(); });` I'd be open to a PR that implements something like that.

**Comment by malaire** [comment]
> It's impossible to both (a) rate limit draws and (b) never show "incorrect" values in case your steps have weirdly distributed timing ...

It's not impossible, just needs separate thread for future draws.

> I definitely don't want to spawn a separate thread to draw in the future

ok, I think I'll just switch to different crate which actually works.

**Comment by djc** [comment]
> It's not impossible, just needs separate thread for future draws.

Fair enough; I feel that the extra complexity is not worth it for a seemingly small amount of use cases. But you should feel free to argue with me if you think I'm wrong on this.

> ok, I think I'll just switch to different crate which actually works.

I'll be sad to see you go, but hope you find something which works better for you.

**Comment by djc** [comment]
FWIW, I rolled back the change that caused this issue on further reflection. We now rely on the draw target's rate limiter, which allows bursts so this issue should not happen.

Also, note that indicatif already had functionality to spawn a background thread that triggers drawing the progress bar at regular intervals.

**Comment by malaire** [comment]
> Also, note that indicatif already had functionality to spawn a background thread that triggers drawing the progress bar at regular intervals.

yes, but that causes unneeded draws when ticker is not being used, i.e. progress bar is redrawn at regular intervals whether it has changed or not.

ps. Anyway I have my own solution now which works better for my use cases.
